src/storage/src/render/multi_worker_persist_sink.rs

```diff
@@ -0,0 +1,1123 @@
+        //
+        // `Antichain` does not implement `Ord`, so we cannot use a `BTreeSet`. We need to search
+        // through the set, so we cannot use the `mz_ore` wrapper either.
+        #[allow(clippy::disallowed_types)]
+        let mut in_flight_descriptions =
+            std::collections::HashSet::<(Antichain<Timestamp>, Antichain<Timestamp>)>::new();
+
+        // In flight batches that haven't been `compare_and_append`'d yet, plus metrics about
+        // the batch.
+        let mut in_flight_batches = HashMap::<
+            (Antichain<Timestamp>, Antichain<Timestamp>),
+            Vec<(Batch<_, _, _, _>, BatchMetrics)>,
+        >::new();
+
+        if !active_worker {
+            // The non-active workers report that they are done snapshotting.
+            source_statistics.initialize_snapshot_committed(&Antichain::<Timestamp>::new());
+            return;
+        }
+
+        // TODO(aljoscha): We need to figure out what to do with error results from these calls.
+        let persist_client = persist_clients
+            .open(persist_location)
+            .await
+            .expect("could not open persist client");
+
+        let mut write = persist_client
+            .open_writer::<SourceData, (), Timestamp, Diff>(
+                shard_id,
+                &format!("persist_sink::append_batches {}", collection_id),
+                Arc::new(target_relation_desc),
+                Arc::new(UnitSchema),
+            )
+            .await
+            .expect("could not open persist shard");
+
+        // Initialize this sink's `upper` to the `upper` of the persist shard we are writing
+        // to. Data from the source not beyond this time will be dropped, as it has already
+        // been persisted.
+        // In the future, sources will avoid passing through data not beyond this upper
+        // VERY IMPORTANT: Only the active write worker must change the
+        // shared upper. All other workers have already cleared this
+        // upper above.
+        current_upper.borrow_mut().clone_from(write.upper());
+        source_statistics.initialize_snapshot_committed(write.upper());
+
+        // The current input frontiers.
+        let mut batch_description_frontier = Antichain::from_elem(TimelyTimestamp::minimum());
+        let mut batches_frontier = Antichain::from_elem(TimelyTimestamp::minimum());
+
+        // Pause the source to prevent committing the snapshot,
+        // if the failpoint is configured
+        let mut pg_snapshot_pause = false;
+        (|| {
+            fail::fail_point!("pg_snapshot_pause", |val| {
+                pg_snapshot_pause = val.map_or(false, |index| {
+                    let index: usize = index.parse().unwrap();
+                    index == output_index
+                });
+            });
+        })();
+
+        loop {
+            tokio::select! {
+                Some(event) = descriptions_input.next_mut() => {
```

Comment:
I don't think this operator needs the description input, does it? Observing the batch stream and its frontiers should have enough data to do the compare and appends which would simplify this operator a lot

Reply:
This is how it works in compute, and this is the least-changed operator between them. I suspect we do it this way because it ensures our batches are grouped the same way they were emitted, which less a correctness requirement, and more a nice-to-have. @aljoscha might know more. This is another example of something that would best be fixed in a later pr, to both sinks, like https://github.com/MaterializeInc/materialize/pull/17589#discussion_r1106016139